account_loan/model/account_loan_line.py

```diff
@@ -307,6 +311,8 @@ def generate_move(self):
                 move = self.env['account.move'].create(record.move_vals())
                 move.post()
                 res.append(move.id)
+        if self.loan_id.state == 'posted':
```

Comment:
I think you should raise a Warning if the loan is not posted before you try to create the move. Moreover, it should be `if rec.loan_id.state != 'posted':` as you could try to create moves from different loans.

Reply:
My add code doesn't create moves it just post a log message for the created moves and after digging and searching for a reason to use the ``` for record in self: ``` loop since all function that call generate_move and generate_invoice gives single record not set of record, I found that generate entries module is giving a set of records and that problem could be solved by adding single loop in the mentioned module instead of writing the loop for each related called function and that will shrink the code and free more resources, anyway I shall include the code that highlighted above to the loop in the next commit to avoid singleton error, also no need for the warning since a warning will be raised if the line wasn't posted and log message will be posted only if the move was created and the loan was posted.
the code will be:
``` python
       for record in self:
            if not record.move_ids:
                if record.loan_id.line_ids.filtered(
                        lambda r: r.date < record.date and not r.move_ids
                ):
                    raise UserError(_('Some moves must be created first'))
                move = self.env['account.move'].create(record.move_vals())
                move.post()
                res.append(move.id)
            if record.move_ids and record.loan_id.state == 'posted':
                record.log_posted_line()
```
thank you for your note.

Reply:
But if you execute the same function twice for the same line, you will post the text twice, so, if is no necessary, and the log_posted_line should be called before move.post.

Reply:
In which case that will happen, this function should execute once for the same line and even if that happend, the message will post only if the move line is created for that loan line and I used ` if record.loan_id.state == 'posted' ` just because the loan close message will appear before the line post message if i remove this part, but if you have some thing that I can't see about why this function will be called twice please give me a clue.

>the log_posted_line should be called before move.post.

if I call log_posted_line before, the message will be posted even if the moves didn't created therefore I need to check for move id before posting the message.